src/types/geography.ts

```diff
@@ -0,0 +1,301 @@
+   */
+  toGeoPoint(): Geography {
+    return new Geography(this.toString());
+  }
+
+  /**
+   * Return new instance of this Geography object
+   * @returns Geography
+   */
+  toGeoLine(): Geography {
+    return new Geography(this.toString());
+  }
+
+  /**
+   * Return new instance of this Geography object
+   * @returns Geography
+   */
+  toGeoPolygon(): Geography {
+    return new Geography(this.toString());
+  }
+
+  /**
+   * Load Geography object represented by the WKT text
+   */
+  fromString(str: string) {
+    this.parse(str);
+  }
+
+  /**
+   * Return the distance (in meters) between the current point and the parameter
+   * @returns number
+   */
+  distance(geo: Geography): number {
+    return Geography.distance(this, geo);
+  }
+
+  /**
+   * Returns true if the Geographic object parameter includes or intersects the current object
+   * @returns boolean
+   */
+  intersect(geo: Geography): boolean {
+    return Geography.intersect(this, geo);
+  }
+
+  /**
+   * Create Geography object represented by the WKT text
+   * @returns Geography
+   */
+  static fromString(str: string): Geography {
+    return new Geography(str);
+  }
+
+  /**
+   * Return the distance (in meters) between points
+   * @returns number
+   */
+  static distance(from: Geography, to: Geography): number {
+    // Haversine formula:
+    // a = sin(delta_lat/2) + cos(lat1).cos(lat2).sin(delta_long/2)
+    // c = 2.atan2(sqrt(a), sqrt(1-a))
+    // d = R.c
+
+    if (from.isEmpty() || to.isEmpty()) return 0;
+
+    const earthRadius = 6371;
+
+    const deltaLatitude = (to.latitude - from.latitude) * (Math.PI / 180);
+    const deltaLongitude = (to.longitude - from.longitude) * (Math.PI / 180);
+
+    const a1 = Math.pow(Math.sin(deltaLatitude / 2), 2);
+    const a2 = Math.pow(Math.sin(deltaLongitude / 2), 2);
+    const a3 =
+      Math.cos(from.latitude * (Math.PI / 180)) *
+      Math.cos(to.latitude * (Math.PI / 180));
+    const a = a1 + a2 * a3;
+
+    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
+
+    return Math.round(earthRadius * c * 1000); // distance in meters
+  }
+
+  /**
+   * Returns true if the two Geographic objects instersect.
+   * @returns boolean
+   */
+  static intersect(a: Geography, b: Geography): boolean {
+    // TODO: not implemented
+    return false;
+  }
+
+  private parse(str: string) {
+    const coordsTestRegex = /^(-?\d*(?:\.\d+)?)(?:\s*,\s*| +)(-?\d*(?:\.\d+)?)$/i; // latitude,longitude
+    const pointTestRegex = /POINT\s*\(\s*(-?\d*(?:\.\d+)?) +(-?\d*(?:\.\d+)?)\s*\)/i; // POINT(longitude latitude)
+    const linestringTestRegex = /LINESTRING *\( *(?:-?\d*(?:\.\d+)? +-?\d*(?:\.\d+)?(?: *, *| *\)$))+/i; // LINESTRING(longitude latitude,longitude latitude)
+    const polygonTestRegex = /POLYGON *\( *(?:\( *(?:-?\d+(?:\.\d+)? +-?\d*(?:\.\d+)?(?: *, *))+-?\d+(?:\.\d+)? +-?\d*(?:\.\d+)? *\)(?: *, *| *\)$))+/i; // POLYGON((longitude latitude,longitude latitude),(longitude latitude,longitude latitude))
```

Comment:
Small performance observation: consider declaring these regexes in constants outside the method, because they will get computed every time the method is called. This could impact performance if many objects are created.

Reply:
I see there are other constant regexes in other lines. Please consider my suggestion for those lines too.

Reply:
What would be better?

class Geography {
...
}
const GEOGRAPHY_REGEX_TEST_COORDS = /^(-?\d*(?:\.\d+)?)(?:\s*,\s*| +)(-?\d*(?:\.\d+)?)$/i; // latitude,longitude

OR

class Geography {
   static GEOGRAPHY_REGEX_TEST_COORDS = /^(-?\d*(?:\.\d+)?)(?:\s*,\s*| +)(-?\d*(?:\.\d+)?)$/i; // latitude,longitude
}

I understand that the first is always processed even if a Geography is not instantiated and the second (in theory) only in the first instance.

Reply:
Yes, that's right. I feel both alternatives look good (considering the class is the only export of the module; if the module is imported, it will probably be because the class will be instantiated). Go with the one you feel the most comfortable about.